internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go

```diff
@@ -83,8 +83,13 @@ func NewServerSidePatchHelper(ctx context.Context, original, modified client.Obj
 	filterObject(modifiedUnstructured, helperOptions)
 
 	// Carry over uid to match the intent to:
-	// * create (uid==""): Setting no uid ensures an object gets created, not updated.
-	// * update (uid!=""): Setting an uid ensures an object which has the same uid gets updated.
+	// * create (uid==""):
+	//   * if object doesn't exist => create
+	//   * if object already exists => update
```

Comment:
Should we also add that we expect that we never hit this case?

Reply:
We hit this case with the clusteShim

Reply:
Do we really?

During reconcileClusterShim: https://github.dev/sbueringer/cluster-api/blob/pr-ssa-helper-doc/internal/controllers/topology/cluster/reconcile_state.go#L104 we always call the patchHelper using `nil` as original which results in `originalUnstructured` to be nil and no UID set, which intents to "create" an object.

Or do we have cases when we enter the PatchHelper for clusterShim, while it exists but we pass `nil`? If that's the case :+1:.

So for all objects we use the ssa patchhelper for, I expect to not hit the `uid==""` + update case.

But ok for me to keep as is.

Reply:
> Or do we have cases when we enter the PatchHelper for clusterShim, while it exists but we pass nil? If that's the case +1.

As you said, we always set original to nil, which means we always try to create the shim (and not set UID). If the line you linked hit is executed multiple times we will have the case that the shim already exists from a previous reconcile.

Before SSA we covered that case by ignoring already exist errors: https://github.com/kubernetes-sigs/cluster-api/blob/release-1.1/internal/controllers/topology/cluster/reconcile_state.go#L94.
Now the SSA will just do an update instead of an create which is then a no-op.

Not sure if I missed something.

Does it make sense for you?

Reply:
Got it 😃